### PR TITLE
implement clean method delete all message of a chat

### LINKED ISSUE #50: Delete whole chat
At the moment, it is possible to delete single messages with Threema Web.

The feature to delete a whole chat would be cool too.

### PR DESCRIPTION
**Pull request by sillych**
Call "Delete Chat" in the receiver detail view to clean a chat (similar to the android app).

2017-09-08-132131_1501x694_scrot

Requires a app update

Fix #50

### CODE REVIEW

File: src/controller_model/contact.ts [hunk 1]
```diff
@@ -111,6 +111,43 @@ export class ContactControllerModel implements threema.ControllerModel {
             );
     }
 
+    public canClean(): boolean {
+        return this.canView();
+            // blocked
+            // && this.contact.;
```
Comment (dbrgn) [inline src/controller_model/contact.ts]: Dead code

File: public/i18n/de.json [hunk 2]
```diff
@@ -150,7 +150,9 @@
         "BALLOT_MESSAGES_NOT_SUPPORTED": "Umfragen werden in Threema Web derzeit nicht untersützt.",
         "UNKNOWN_MESSAGE_TYPE": "Unbekannter Nachrichtentyp",
         "NICKNAME": "Nickname",
-        "THREEMA_WORK_CONTACT": "Threema Work Nutzer"
+        "THREEMA_WORK_CONTACT": "Threema Work Nutzer",
+        "DELETE_THREAD": "Chat löschen",
+        "DELETE_THREAD_MESSAGE": "{count, plural, one {Möchten Sie wirklich diesen Chat löschen? Die Nachrichten können nicht wiederhergestellt werden.} other {Möchten Sie wirklich # Chats löschen? Die Nachrichten können nicht wiederhergestellt werden.}}"
```
Comment (dbrgn) [inline public/i18n/de.json]: Do you plan to support deleting multiple chats at once? So far we only use count 1. If not, I'd remove the plural forms.

File: src/controller_model/contact.ts [hunk 3]
```diff
@@ -111,6 +111,43 @@ export class ContactControllerModel implements threema.ControllerModel {
...
+    }
+
+    private reallyClean(): any {
+        if (!this.canClean()) {
+            this.$log.error('not allowed to clean this contact');
+            return;
+        }
+
+        this.isLoading = true;
+        this.webClientService.clean(this.contact)
```
Comment (dbrgn) [inline src/controller_model/contact.ts]: Maybe the `clean` method should be renamed to `cleanChat` or `deleteChat` to make it easier to understand what this does.

File: src/services/webclient.ts [hunk 4]
```diff
@@ -1227,6 +1228,24 @@ export class WebClientService {
     }
 
     /**
+     * Remove all messages of a receiver
+     * @param {threema.Receiver} receiver
+     * @returns {Promise<any>}
+     */
+    public clean(receiver: threema.Receiver): Promise<any> {
+        if (receiver === undefined) {
```
Comment (dbrgn) [inline src/services/webclient.ts]: We should probably check for `null` too. (The joys of Javascript.)

Comment (sillych) [reply src/services/webclient.ts]: feel free to append this check :laughing:

File: public/i18n/de.json (hunk 2 above)
Comment (sillych) [reply public/i18n/de.json]: yes is planned...

**Comment by sillych** [comment]
- If the App does not support this feature, nothing happens, i think its ok.
- After cleaning a conversation, the cache (stored messages on browser) will be cleared now
- Chats will be closed now (app) after a clean (webclient)

**Pull request merged**